### ISSUE TITLE
logsigbackprop gives incorrect values

### ISSUE THREAD

**Issue by patrick-kidger** [bottler/iisignature#8]
So this is a subtle one, but I'm as sure as I can be that it's present.

In brief, `logsigbackprop` gives values which are slightly off (by about ~10^-3 in my typical test cases). This issue only appears to manifest when the alphabet size is at least 3, the length of the input is at least 3, and the depth is at least 4.

I'm using PyTorch to aid in some autograd-derived backwards computations.

```python
import torch
import iisignature as ii

# Assumptions: (a) ii.logsig(...) is correct
#              (b) ii.logsig(... ,'x') is correct
#              (c) ii.logsigbackprop(..., 'x') is correct
#              (d) PyTorch does backprop correctly

# Result: ii.logsigbackprop(...) is incorrect.

# Method: Call ii.logsig(..., 'x'). Compute the 'compressed' (so to speak) logsignature from this
# using PyTorch. We verify that the implementation is correct by comparing against
# ii.logsig(...).
# Then compute the backwards through these operations using PyTorch, then the backwards
# through the logsignature with ii.logsigbackprop(..., 'x').
# We find that the value thus derived is not equal to the value given by ii.logsigbackprop(...).

# alphabet size = 3, depth = 4
# These appear to be the minimal values for which this discrepancy occurs
path = torch.rand(3, 3, dtype=torch.double)
logsig_expanded = ii.logsig(path, ii.prepare(3, 4), 'x')
logsig_expanded = torch.tensor(logsig_expanded, requires_grad=True)

# Indices corresponding to Lyndon words
indices = [0, 1, 2, 4, 5, 8, 13, 14, 16, 17, 19, 20, 26, 29, 40, 41, 43, 44, 46, 47, 50, 52, 53,
           55, 56, 61, 62, 64, 65, 80, 83, 92]
# These describe the linear transformations necessary to get into the Lyndon basis
transforms = [(9, 10, -1), (17, 18, -1), (18, 20, -1), (22, 23, -2), (22, 25, 1), (23, 25, -1),
              (24, 26, -2), (24, 27, 1), (26, 27, -1)]

logsig_transform_old = torch.stack([logsig_expanded[i] for i in indices])
for source, target, coefficient in transforms:
    logsig_transform = logsig_transform_old.clone()
    logsig_transform[target] -= logsig_transform_old[source] * coefficient
    logsig_transform_old = logsig_transform

# Verify that our PyTorch-based forward computation is correct (Really, the mathematical
# derivation of our above calculations should be performed, or this following check performed
# many times. I've done that, just not in this example; I'll let you check.)
logsig_lyndon = torch.tensor(ii.logsig(path, ii.prepare(3, 4)))
print('Difference in forward: ', torch.max(torch.abs(logsig_lyndon - logsig_transform)))

# Now perform the PyTorch-based based backward computation.
z = torch.rand_like(logsig_transform)
logsig_transform.backward(z)
correct_backward = ii.logsigbackprop(logsig_expanded.grad, path, ii.prepare(3, 4), 'x')
incorrect_backward = ii.logsigbackprop(z, path, ii.prepare(3, 4))
print('Difference in backward: ',
      torch.max(torch.abs(torch.tensor(correct_backward - incorrect_backward))))
```

**Comment by bottler** [comment]
I think I agree there's a problem with this function with the Lyndon basis. I can observe it with the iisignature tests themselves: if I change the dimension to 3 in Deriv.logSig function in test_sig.py then the relevant test case fails. It may take a while to sort out.

**Comment by bottler** [comment]
I think what I've just committed should fix it. Should now be ok if you install from github. 

https://github.com/bottler/iisignature/commit/7ad9af

**Comment by bottler** [comment]
Recently released version 0.24 has this fixed.
